### PR TITLE
mount: Reduce readings during ec write

### PR DESCRIPTION
**Pull request by rolysr**
Tests performed show that during EC writes the client connects and asks
to the chunkservers for the missing data parts of the current stripe,
which are the data parts the client are not writing. This is required if
the file is bigger than the last byte it is writing, but it should not
ask for data on offsets beyond the file size. It is specially important
for the writings of small files, and removing those extra requests
should reduce overhead.

Current solution manages to calculate chunkSizeInBlocks_ for each
ChunkWriter at the moment of initialization from the field maxfleng of
the current inode's inodedata. Thus, the maxfleng field should be
updated the best way possible. The chunkSizeInBlocks_ is then used to
check if client ws about to was to chunkservers for not existing data.

### CODE REVIEW

**Review by aNeutrino — CHANGES_REQUESTED** [review]
Missing 2..3 automated test suits which would test new functionality.

**Review by uristdwarf — COMMENTED** [review]
I'm not very familiar in this area, so I won't approve or request changes specifically for approval, but I do have a few comments.

1. If it's possible, try to make a unit test for `fillNotExisting`. I say if possible because I understand there may be certain situations it's not because of external dependencies. But on first glance I would try, since we really lack unit tests and this is a good candidate.
2. This will probably come in a later release (4.2 most likely), so this probably won't be merged before Wednesday.

**Comment by uristdwarf** [comment]
Short System tests `test_auto_recovery_incversion`, `test_auto_recovery_nextchunkid` and `test_multiple_master_promotions` are failing on my system, compared to dev

File: src/mount/sauna_client.cc [hunk 1]
```diff
@@ -2359,7 +2359,18 @@ BytesWritten write(Context &ctx, Inode ino, const char *buf, size_t size, off_t
...
 		fileinfo->data = write_data_new(ino);
 	}
-	err = write_data(fileinfo->data,off,size,(const uint8_t*)buf);
+
+	Attributes attr;
+	attr.fill(0);
+	if (usedircache) { 
+		gDirEntryCache.lookup(ctx, ino, attr); 
+	}
+	stat *stbuf = new stat;
```
[the remark this answers is not in the document]
Comment (uristdwarf) [reply src/mount/sauna_client.cc]: Ah dangit, didn't see it was fixed in the next commit in the series. Sorry

File: src/chunkserver/hddspacemgr.cc [hunk 2]
```diff
@@ -775,6 +775,8 @@ int hddRead(uint64_t chunkId, uint32_t version, ChunkPartType chunkType,
 	LOG_AVG_TILL_END_OF_SCOPE0("hddRead");
 	TRACETHIS3(chunkId, offset, size);
 
+	safs_silent_syslog(LOG_DEBUG, "chunkserver.hddRead");
```
Comment (dmga44) [inline src/chunkserver/hddspacemgr.cc:778]: Is this line required by the test?

**Comment by dmga44** [comment]
> @rolysr do we have test which covers this case? Eg. 2 separate mount points Create file and Set gooal EC(4,2)
> 
> write 64KiB...128KiB byte on byte 0 on 1st mount point (part 1 and/or 2) In parallel (same time) 2nd mount point should write 1byte/64KiB to position 192KiB (part 4 only) or position 128KiB (part 3 only)
> 
> Above is example of 6 test cases Which must check that one part is(3 or 4) never read and final file include all data from parts 1,2 and (3 or 4).
> 
> I do not see any new tests in your PR...it seems wrong.
> 
> I would like to se 2 or 3 test suits as proposed above.

I would like to comment a bit here.

To my understanding, you are asking for a test which after creating a file and setting its goal, performs in parallel the following two operations:
- write a number of bytes from 64KiB to 128KiB to the file starting from offset 0 (meaning writes on blocks 1 and/or 2)
- write a number of bytes from 1 to 64KiB (it doesn't really matter which) to position 192KiB (meaning writes on the block 4) or equivalent for block 3

So, I think we have two scenarios:
- the mountpoint 1 is the first one to lock the chunk and performs its writes (block 1 and/or 2) without reading anything extra from the chunkservers, and calculates the parity parts. The mountpoint 2 then performs its writes (block 4), but to calculate the parity parts, it needs to read from the previosly written blocks from the first mountpoint.
- the mountpoint 2 is the first one to lock the chunk and performs its writes (block 4) and to calculate the parity parts it needs to read the first 3 blocks, which should be full of zeroes. The mountpoint 1 then performs its writes (blocks 1 and/or 2), and to calculate the parity parts, it needs to read from the previosly existing blocks from the stripe (blocks 3 (full of zeroes) and 4 (some content)).

Anyway we'll see some readings to calculate parities.

File: src/chunkserver/hddspacemgr.cc (hunk 2 above)
Comment (rolysr) [reply src/chunkserver/hddspacemgr.cc:778]: Yes, that line is used for checking the extra read operations

File: tests/test_suites/ShortSystemTests/test_ec_truncate_atomicity.sh [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+done
+# In the meanwhile, shorten file from 400K to 200K, chopping 1 kB in each of 200 steps
+# Sometimes share mountpoint with dd (mount2), sometimes not (mount3, mount4)
+for i in {399..200}; do
+	truncate -s ${i}K "${info[mount$((2 + i % 3))]}/$file"
+done
+wait # Wait for all dd processes to finish
+
+# Now file should be equal to the source file. Let's validate it!
+MESSAGE="Testing ec(4,2) config" assert_success file-validate "$file"
```
Comment (lgsilva3087) [inline tests/test_suites/ShortSystemTests/test_ec_truncate_atomicity.sh:36]: This tests is similar to test_xor_truncate_atomicity. In the original test, the first stripe was removed in this point to check if the system can read also from the parities. Please, consider doing the same.

File: tests/test_suites/ShortSystemTests/test_reads_during_small_ec_files_write_alternated.sh [hunk 4]
```diff
@@ -0,0 +1,39 @@
+timeout_set 120 seconds
+
+CHUNKSERVERS=6 \
+	MOUNTS=2 \
+	USE_ZONED_DISKS=YES \
```
Comment (lgsilva3087) [inline tests/test_suites/ShortSystemTests/test_reads_during_small_ec_files_write_alternated.sh:5]: Same here. This repo should not contain specific SMR tests.

File: tests/test_suites/ShortSystemTests/test_reads_during_small_ec_files_write_alternated.sh [hunk 5]
```diff
@@ -0,0 +1,39 @@
+timeout_set 120 seconds
```
Comment (lgsilva3087) [inline tests/test_suites/ShortSystemTests/test_reads_during_small_ec_files_write_alternated.sh:1]: How much time is consuming this test? 120 seconds is a long time.

**Review by lgsilva3087 — CHANGES_REQUESTED** [review]
Please see my comments in the tests.
This change was originated because there was a lot of reading during creation of small files in EC. Please consider adding a test that creates 100 small EC files (33 bytes) and check that there are no unneeded reads.

File: src/mount/chunk_locator.h [hunk 6]
```diff
@@ -97,6 +97,10 @@ class WriteChunkLocator {
 	virtual void locateAndLockChunk(uint32_t inode, uint32_t index);
 	virtual void unlockChunk();
 
+	uint64_t fileLength() const {
```
[the remark this answers is not in the document]
Comment (rolysr) [reply src/mount/chunk_locator.h]: Done

File: src/mount/chunk_writer.cc [hunk 7]
```diff
@@ -328,6 +328,11 @@ std::list<WriteCacheBlock> ChunkWriter::releaseJournal() {
 	return std::move(journal_);
 }
 
+void ChunkWriter::setChunkSizeInBlocks(uint32_t chunkSize) {
```
[the remark this answers is not in the document]
Comment (rolysr) [reply src/mount/chunk_writer.cc]: Done

File: src/mount/chunk_writer.cc [hunk 8]
```diff
@@ -436,6 +441,46 @@ void ChunkWriter::fillOperation(Operation &operation, int first_block, int first
...
 
+/*!
+ * Fills given operation with a range of blocks beyond file size required to be read.
+ * \param operation operation to be filled
+ * \param first_block first block of the operation
+ * \param first_index index of the first block required to be read
+ * \param size number of required blocks
+ * \param stripe_element map of blocks in a stripe
+ */
+void ChunkWriter::fillNotExisting(Operation &operation, int first_block, int first_index, int size,
```
[the remark this answers is not in the document]
Comment (rolysr) [reply src/mount/chunk_writer.cc]: Done

File: src/mount/chunk_writer.cc [hunk 9]
```diff
@@ -436,6 +441,46 @@ void ChunkWriter::fillOperation(Operation &operation, int first_block, int first
...
+		std::vector<uint8_t *> &stripe_element) {
+	assert(size >= 0);
```
[the remark this answers is not in the document]
Comment (rolysr) [reply src/mount/chunk_writer.cc]: There is no code calling that function with 0 blocks.

File: src/mount/chunk_writer.h [hunk 10]
```diff
@@ -127,6 +127,8 @@ class ChunkWriter {
 
 	bool acceptsNewOperations() { return acceptsNewOperations_; }
 
+	void setChunkSizeInBlocks(uint32_t chunkSize);
```
[the remark this answers is not in the document]
Comment (rolysr) [reply src/mount/chunk_writer.h]: Done

File: tests/test_suites/ShortSystemTests/test_ec_truncate_atomicity.sh (hunk 3 above)
Comment (rolysr) [reply tests/test_suites/ShortSystemTests/test_ec_truncate_atomicity.sh]: Done

File: tests/test_suites/ShortSystemTests/test_reads_during_small_ec_files_write.sh [hunk 11]
```diff
@@ -0,0 +1,85 @@
+timeout_set 120 seconds
+
+CHUNKSERVERS=6 \
+	MOUNTS=2 \
+	USE_ZONED_DISKS=YES \
```
[the remark this answers is not in the document]
Comment (rolysr) [reply tests/test_suites/ShortSystemTests/test_reads_during_small_ec_files_write.sh]: Done

File: tests/test_suites/ShortSystemTests/test_reads_during_small_ec_files_write_alternated.sh (hunk 5 above)
Comment (rolysr) [reply tests/test_suites/ShortSystemTests/test_reads_during_small_ec_files_write_alternated.sh]: Done

File: tests/test_suites/WindowsClientTests
Comment (lgsilva3087) [inline tests/test_suites/WindowsClientTests:1]: This soft link should not be here. Please verify.

**Review by lgsilva3087 — CHANGES_REQUESTED** [review]
Looks good. Please verify the soft link.

**Review by dmga44 — COMMENTED** [review]
LGTM, though I'm coauthor.

File: src/mount/chunk_writer.cc [hunk 12]
```diff
@@ -436,6 +436,48 @@ void ChunkWriter::fillOperation(Operation &operation, int first_block, int first
...
 
+/*!
+ * Fills given operation with a range of blocks beyond file size required to be read.
+ * \param operation operation to be filled
+ * \param first_block first block of the operation
+ * \param first_index index of the first block required to be read
+ * \param size number of required blocks
+ * \param stripe_element map of blocks in a stripe
+ */
+void ChunkWriter::fillNotExisting(Operation &operation, int first_block,
```
Comment (Antuan-AE) [inline src/mount/chunk_writer.cc:447]: Consider declaring first_block parameter as unsigned integer.

File: src/mount/chunk_writer.h [hunk 13]
```diff
@@ -175,6 +180,7 @@ class ChunkWriter {
 	bool acceptsNewOperations_;
 	int combinedStripeSize_;
 	int dataChainFd_;
+	int chunkSizeInBlocks_;
```
Comment (Antuan-AE) [inline src/mount/chunk_writer.h:183]: Consider declaring this variable as unsigned interger as negative values have not practical meaning for this variable.

File: src/mount/sauna_client.cc [hunk 14]
```diff
@@ -2359,7 +2359,17 @@ BytesWritten write(Context &ctx, Inode ino, const char *buf, size_t size, off_t
...
+	Attributes attr;
+	attr.fill(0);
+	if (usedircache) {
+		gDirEntryCache.lookup(ctx, ino, attr); 
+	}
+	struct stat stbuf;
+	attr_to_stat(ino, attr, &stbuf);
+	size_t currentSize = stbuf.st_size;
+
+	err = write_data(fileinfo->data,off,size,(const uint8_t*)buf, currentSize);
```
Comment (Antuan-AE) [inline src/mount/sauna_client.cc:2372]: Consider adding a space between comma separated parameters to comply with code style.

**Review by Antuan-AE — APPROVED** [review]
LGTM

File: src/mount/chunk_writer.cc (hunk 12 above)
Comment (lgsilva3087) [reply src/mount/chunk_writer.cc:447]: We will leave it for now to match with already existing code.
image

File: src/mount/chunk_writer.h (hunk 13 above)
Comment (lgsilva3087) [reply src/mount/chunk_writer.h:183]: Same here.

File: src/mount/sauna_client.cc (hunk 14 above)
Comment (lgsilva3087) [reply src/mount/sauna_client.cc]: Done.

**Comment by lgsilva3087** [comment]
@aNeutrino your requested changes seem to be addressed with the newly added tests. Please re-check.

**Review by aNeutrino — APPROVED** [review]
LGTM

**Pull request merged**